File: tools/surface_match.mjs
// 2026-09-10: WHICH SURFACE AGREES WITH THE PHOTOGRAPH, NOT WHICH PICTURE. THE IDENTITY PASS.
//
// tools/render_match.py put the sim camera where a real camera stood and scored the whole frame. That can
// only ever say "this picture agrees" or "this picture does not", and a hall frame is mostly ceiling and
// floor, so a balcony fascia drawn in the wrong place is a per cent of the score and invisible inside it.
//
// SO SHOOT THE SAME POSE TWICE. Once normally, and once with the surfaces this goal names replaced by a
// flat identity colour, one colour per NAME, everything else painted black. The second pass is not a
// picture of the hall, it is a map: each pixel says which named surface the model drew there. Scoring the
// photograph inside one surface's mask asks whether THAT surface is where the photograph puts it, and the
// answer cannot be diluted by the ceiling.
//
// THE COLOURS ARE FOUR LEVELS A CHANNEL AND NOTHING FINER, because the screenshot comes back as JPEG and
// a fine grid would let compression slide one surface into its neighbour. Eighty-five levels between
// neighbours against a tolerance of twenty-eight is the margin, and only the balcony, wall and corridor
// surfaces get a colour at all. The scorer prints what fraction of the
// picture it could classify; if that is poor, nothing downstream is worth reading and the number is shown
// rather than hidden. NOTHING HERE CHANGES THE PAGE ON DISK: the materials are swapped in the live browser
// and put back before it closes, and the run says so at the end.
//   bash ~/scripts/headless-chrome.sh start 9334
//   CDP_PORT=9334 node tools/surface_match.mjs
//   bash ~/scripts/headless-chrome.sh stop 9334
import fs from 'node:fs';
import { attach } from './cdp.mjs';

const PAGE = process.env.SHOT_URL || 'https://lotron-electrical.github.io/ngv-hall-model/index.html';
const picks = JSON.parse(fs.readFileSync('render-match.json', 'utf8'));
const RH = 720;
const P = await attach({ port: +(process.env.CDP_PORT || 9334), width: 1280, height: RH });
await P.send('Network.setCacheDisabled', { cacheDisabled: true }).catch(() => {});
console.log('loading', PAGE);
await P.send('Page.navigate', { url: `${PAGE}?install=gandel-2026&cb=${Date.now()}` });
let ok = false;
for (let i = 0; i < 150; i++) { await P.sleep(500); if (await P.ev('!!(window.ngv&&window.ngv.hall&&document.getElementById("install"))').catch(() => false)) { ok = true; break; } }
if (!ok) { console.log('THE PAGE NEVER REACHED INSTALL MODE. Nothing can be rendered.'); await P.close(); process.exit(1); }
await P.ev(`localStorage.setItem('ngv.install','gandel-2026'); document.getElementById('install').click()`);
for (let i = 0; i < 60; i++) { await P.sleep(500); if (await P.ev('!!(window.ngv&&window.ngv.game)').catch(() => false)) break; }
await P.ev(`document.querySelector('#start').click()`).catch(() => {});
await P.sleep(1200);
await P.ev(`(()=>{document.body.classList.add("immersive");
  const cb=document.getElementById("collide"); if(cb){cb.checked=false; cb.dispatchEvent(new Event("change"));}
  const keep=document.getElementById('cv');
  document.querySelectorAll('body > *, main > *, #stage > *').forEach(e=>{ if(e===keep||e.contains(keep))return; e.style.visibility='hidden'; });
  let n=0; ngv.scene.traverse(o=>{ if(o.isSprite){o.visible=false;n++;} }); ngv.dirty(); return n;})()`)
  .then(n => console.log('sprites hidden', n)).catch(e => console.log('hide failed', String(e.message).slice(0, 120)));

// WHICH SURFACES GET A COLOUR. The goal names three things, so the filter names three things, and the
// list of what matched is printed so nobody has to trust the pattern. Everything else goes black, which
// also makes the background unambiguous.
const WANT = 'gallery|corridor|wall|rail|fascia|parapet|upstand|downstand|soffit|deck|reveal|jamb|opening|pier|tier|balcon|brick';
const table = await P.ev(`(()=>{ const T=window.dbg&&window.dbg.THREE; if(!T) return {err:'no THREE handle'};
  const L=[0,85,170,255]; const pool=[];
  for(const r of L) for(const g of L) for(const b of L){ if(r+g+b<85) continue; pool.push([r,g,b]); }
  const re=new RegExp('${WANT}','i'); const names=[]; const others=[];
  ngv.scene.traverse(o=>{ if(!o.isMesh) return; const nm=o.name||(o.material&&o.material.name)||'(unnamed)';
    if(re.test(nm)){ if(!names.includes(nm)) names.push(nm); } else if(!others.includes(nm)) others.push(nm); });
  window.__idmap={}; window.__saved=[];
  names.slice(0,pool.length).forEach((nm,i)=>{ window.__idmap[nm]=pool[i]; });
  return {names, colours: names.map(n=>window.__idmap[n]||null), pool:pool.length, others:others.length}; })()`)
  .catch(e => ({ err: String(e.message).slice(0, 200) }));
if (table.err) { console.log('the identity pass cannot be built:', table.err); await P.close(); process.exit(1); }
console.log(`${table.names.length} named surfaces match the goal, ${table.pool} colours available, ${table.others} other names go black`);
if (table.names.length > table.pool) console.log(`ONLY THE FIRST ${table.pool} GET A COLOUR, the rest are black and unscored: ${table.names.slice(table.pool).join(', ')}`);
console.log('   ' + table.names.slice(0, table.pool).join(', '));
fs.writeFileSync('render-match/idmap.json', JSON.stringify(table, null, 1));

// MODE 'bounds': EXPORT WHERE EACH NAMED SURFACE ACTUALLY IS, then stop. The first run of this tool
// scored ten frames chosen for how much OPENING they showed, and discovered that the balcony surfaces
// never fill a readable share of any of them: gallery-fascia, gallery-rail, gallery-floor and the rest
// returned zero pixels in all ten. That is not a fault in the scoring, it is the frame set being aimed at
// the wrong thing, and it cannot be fixed by looking at pictures. So each mesh gives up its own world
// bounding box here, in hall coordinates, and the frames get chosen per SURFACE from those boxes.
if (process.argv[2] === 'bounds') {
  const O = [-54.907447, -1.43545, 3.040286];
  const HU = [0.975681, 0, 0.219196];
  const HD = [0.219196, 0, -0.975681];
  const box = await P.ev(`(()=>{ const O=${JSON.stringify(O)}, HU=${JSON.stringify(HU)}, HD=${JSON.stringify(HD)};
    const toHall = w => { const q=[w[0]-O[0], w[1]-O[1], w[2]-O[2]];
      return [q[0]*HU[0]+q[2]*HU[2], q[0]*HD[0]+q[2]*HD[2], q[1]]; };
    const out={};
    ngv.scene.traverse(o=>{ if(!o.isMesh||!o.geometry) return;
      const nm=o.name||(o.material&&o.material.name)||'(unnamed)';
      const pos=o.geometry.attributes&&o.geometry.attributes.position; if(!pos) return;
      o.geometry.computeBoundingBox(); const bb=o.geometry.boundingBox; if(!bb) return;
      o.updateWorldMatrix(true,false); const M=o.matrixWorld.elements;
      const put=(x,y,z)=>{ const w=M[3]*x+M[7]*y+M[11]*z+M[15];
        return toHall([(M[0]*x+M[4]*y+M[8]*z+M[12])/w,(M[1]*x+M[5]*y+M[9]*z+M[13])/w,(M[2]*x+M[6]*y+M[10]*z+M[14])/w]); };
      // ONE BOX PER PART, NOT ONE PER NAME. gallery-fascia is two meshes forty metres apart, one on each
      // end gallery, and a box round both of them is a box round the whole hall: no camera can frame it
      // and the surface would never be selectable. Grouping back by name happens in the scorer.
      const e=out[nm]||(out[nm]={parts:[]});
      const mn=[1e9,1e9,1e9], mx=[-1e9,-1e9,-1e9];
      for(const x of [bb.min.x,bb.max.x]) for(const y of [bb.min.y,bb.max.y]) for(const z of [bb.min.z,bb.max.z]){
        const q=put(x,y,z); for(let k=0;k<3;k++){ if(q[k]<mn[k])mn[k]=q[k]; if(q[k]>mx[k])mx[k]=q[k]; } }
      e.parts.push({min:mn, max:mx, verts:pos.count}); });
    return out; })()`).catch(e => ({ err: String(e.message).slice(0, 200) }));
  if (box.err) { console.log('the bounds export failed:', box.err); await P.close(); process.exit(1); }
  const names = Object.keys(box);
  for (const nm of names) box[nm].baked = box[nm].parts.some(t => t.verts > 4000);
  fs.writeFileSync('render-match/bounds.json', JSON.stringify(box, null, 1));
  const np = names.reduce((s, n) => s + box[n].parts.length, 0);
  console.log(`${names.length} named meshes over ${np} parts, boxes written to render-match/bounds.json`);
  console.log(`${names.filter(n => box[n].baked).length} of them are baked scan pieces and are marked as such`);
  await P.close();
  process.exit(0);
}

// AND THE PASS THAT TELLS YOU WHY A SURFACE HAS NO PIXELS. The first scored run found every goal surface
// drawing a thousandth of the picture its own bounding box fills, which is either a thin surface seen
// edge-on or a surface hidden behind something else, and those are very different facts.
// THE FIRST X-RAY WAS BUILT WRONG AND ITS OWN OUTPUT CAUGHT IT. Turning DEPTH TESTING OFF let the goal
// surfaces draw over each other in traversal order, so end-ground-wall came back showing 937k pixels
// normally and 266k with nothing supposed to be covering it: a surface cannot be more visible than its own
// silhouette, and a negative hidden fraction is a broken instrument, not a finding.
// THE RIGHT X-RAY HIDES EVERYTHING THAT IS NOT A GOAL SURFACE and leaves depth testing alone. The goal
// surfaces then occlude each other exactly as they really do, nothing else can cover them, and the
// shortfall against the ordinary pass is what the rest of the model hides. Measured, not assumed.
const XRAY = process.argv[2] === 'xray';
const idOn = () => P.ev(`(()=>{ const T=dbg.THREE; window.__saved=[]; window.__bg=ngv.scene.background;
  ngv.scene.background=new T.Color(0,0,0); window.__fog=ngv.scene.fog; ngv.scene.fog=null;
  ngv.scene.traverse(o=>{ if(!o.isMesh) return; const nm=o.name||(o.material&&o.material.name)||'(unnamed)';
    const c=window.__idmap[nm]; window.__saved.push([o,o.material,o.visible]);
    const m=new T.MeshBasicMaterial({side:T.DoubleSide, fog:false, toneMapped:false});
    m.color.setRGB((c?c[0]:0)/255,(c?c[1]:0)/255,(c?c[2]:0)/255); o.material=m;
    if(${XRAY} && !c) o.visible=false; });
  ngv.dirty(); return window.__saved.length; })()`).catch(e => String(e.message).slice(0, 160));
const idOff = () => P.ev(`(()=>{ for(const [o,m,v] of (window.__saved||[])){ o.material=m; if(v!==undefined)o.visible=v; } window.__saved=[];
  if(window.__bg!==undefined) ngv.scene.background=window.__bg; if(window.__fog!==undefined) ngv.scene.fog=window.__fog;
  ngv.dirty(); return 'restored'; })()`).catch(() => 'restore failed');

for (let i = 0; i < picks.length; i++) {
  const p = picks[i];
  const side = 2 * Math.round(RH * p.sqside / 2);
  await P.send('Emulation.setDeviceMetricsOverride', { width: side, height: side, deviceScaleFactor: 1, mobile: false });
  await P.ev(`window.dispatchEvent(new Event('resize'))`).catch(() => {});
  await P.sleep(700);
  const hour = p.cls === 'night' ? 21 : 13;
  const house = p.cls === 'night' ? 70 : 0;
  await P.ev(`(()=>{const t=document.getElementById('tod'); t.value=${hour}; t.dispatchEvent(new Event('input'));
    const hs=document.getElementById('house'); if(hs){hs.value=${house}; hs.dispatchEvent(new Event('input'));} ngv.dirty();})()`).catch(() => {});
  await P.sleep(700);
  const place = () => P.ev(`(()=>{const g=ngv.game,PL=g.player,L=g.lift; const a=g.hallToWorld(${p.u},${p.d},g.world.floorY);
    const b=g.hallToWorld(${p.u + p.fu},${p.d + p.fd},g.world.floorY); PL.yaw=Math.atan2(-(b.x-a.x),-(b.z-a.z)); PL.pitch=${p.pitch}*Math.PI/180;
    if(L.aboard)L.leave(PL,true); PL.pos.copy(a); PL.pos.y=g.world.floorY; PL.eye=${p.h}; return PL.eye;})()`).catch(e => String(e.message).slice(0, 120));
  await place();
  await P.sleep(1000);
  await place();
  await P.ev(`(()=>{const g=ngv.game; if(g.fx)g.fx.baseFov=${p.sqvfov}; ngv.cam.fov=${p.sqvfov}; ngv.cam.updateProjectionMatrix(); return ngv.cam.fov;})()`).catch(() => {});
  await place();
  await P.sleep(400);
  await P.shot(`render-match/r${String(i).padStart(2, '0')}.jpg`);
  const n = await idOn();
  await P.sleep(900);
  await place();
  await P.sleep(500);
  await P.shot(`render-match/${XRAY ? 'xr' : 'id'}${String(i).padStart(2, '0')}.jpg`);
  await idOff();
  await P.sleep(400);
  console.log(`${p.cls} ${p.frame}  square ${side}  ${n} meshes recoloured  -> r${String(i).padStart(2, '0')}.jpg + id${String(i).padStart(2, '0')}.jpg`);
}
console.log('');
console.log('materials restored:', await idOff());
console.log(P.errors.length ? `the page reported ${P.errors.length} errors: ${P.errors.slice(0, 4).join(' | ')}`
  : 'the page reported no errors while any of this ran.');
await P.close();
process.exit(0);

File: tools/surface_match.py
# 2026-09-10: SCORE THE PHOTOGRAPH INSIDE EACH SURFACE'S OWN MASK.
#
# THE QUESTION THE WHOLE-FRAME TEST COULD NOT ASK. tools/render_match.py showed the sim lands on the
# photograph well enough to beat every mismatched pair, and gave a bound of about a degree. But a hall
# frame is mostly ceiling and floor, so a balcony fascia in the wrong place moves that score by a per cent
# and disappears into it. This scores each named surface separately, inside the pixels the model itself
# says that surface covers, which comes from the identity pass tools/surface_match.mjs renders.
#
# THE DECISION RULE, FIXED HERE BEFORE THE RUN.
#   1. A surface is only read where the model draws at least MINPIX of it in a frame, after the mask is
#      eroded so that no antialiased edge pixel is counted.
#   2. Its score is the normalised cross-correlation of the gradients inside that mask, taken where a
#      perfect model would put the render, which is the principal point offset and not zero.
#   3. THE CONTROL IS THE SAME MASK ON A DIFFERENT FRAME'S PHOTOGRAPH. That answers the objection that a
#      mask over a busy part of the picture scores well no matter what is drawn there.
#   4. A surface is CORROBORATED in a frame when its true score beats every control score for that same
#      mask. Anything else is NOT CORROBORATED, which is not the same as refuted and is never written as
#      though it were.
#   5. A surface the model draws but no frame can corroborate is reported as DRAWN ON NOTHING VISIBLE,
#      which is a statement about the archive, not about the building.
#
# WHAT IT CANNOT DO. The mask comes from the model, so a surface drawn in completely the wrong place is
# scored against whatever the photograph happens to show there, and will fail rather than point anywhere.
# It cannot separate a surface being wrong from the camera pose being wrong. And much of what it scores is
# the baked scan, which came from photographs of this building, so a high score there is partly the bake
# agreeing with its own source; the baked names are marked in the output.
#   python tools/surface_match.py
import io
import json
import os
import sys

import cv2
import numpy as np

OUT = 'render-match'
WH = 540
MINPIX = 2500
ERODE = 5

sys.path.insert(0, 'tools')


def grad(im):
    g = cv2.GaussianBlur(im.astype(np.float32), (0, 0), 1.6)
    m = np.hypot(cv2.Sobel(g, cv2.CV_32F, 1, 0, ksize=3), cv2.Sobel(g, cv2.CV_32F, 0, 1, ksize=3))
    return m / max(float(m.std()), 1e-6)


def turn(p, im, nearest=False):
    """Put the square render on its side by the camera's own roll, then cut the frame out of the middle."""
    side = im.shape[0]
    c = (side / 2.0 - 0.5, side / 2.0 - 0.5)
    want = np.array([p['upx'], p['upy']])
    best, bM = None, None
    for a in (p['roll'], -p['roll']):
        M = cv2.getRotationMatrix2D(c, a, 1.0)
        e = float(np.hypot(*(M[:2, :2] @ np.array([0.0, -1.0]) - want)))
        if best is None or e < best:
            best, bM = e, M
    fl = cv2.INTER_NEAREST if nearest else cv2.INTER_AREA
    im = cv2.warpAffine(im, bM, (side, side), flags=fl, borderValue=0)
    kw = int(round(side * p['w'] / float(p['hgt']) / p['sqside']))
    kh = int(round(side / p['sqside']))
    x0, y0 = (side - kw) // 2, (side - kh) // 2
    wide = int(round(WH * p['w'] / float(p['hgt'])))
    return cv2.resize(im[y0:y0 + kh, x0:x0 + kw], (wide, WH), interpolation=fl)


def photo(p):
    im = cv2.imread(p['photo'], cv2.IMREAD_GRAYSCALE)
    if im is None:
        return None
    q = p['params']
    if len(q) > 4:
        K = np.array([[q[0], 0, q[2]], [0, q[1], q[3]], [0, 0, 1.0]])
        im = cv2.undistort(im, K, np.array(q[4:8], float), None, K)
    wide = int(round(WH * p['w'] / float(p['hgt'])))
    return grad(cv2.resize(im, (wide, WH), interpolation=cv2.INTER_AREA))


def masked_ncc(A, B, mask, dx, dy):
    h, w = A.shape
    x0, x1 = max(0, dx), min(w, w + dx)
    y0, y1 = max(0, dy), min(h, h + dy)
    a = A[y0:y1, x0:x1]
    b = B[y0 - dy:y1 - dy, x0 - dx:x1 - dx]
    m = mask[y0:y1, x0:x1]
    if m.sum() < MINPIX:
        return None
    av, bv = a[m], b[m]
    av = av - av.mean()
    bv = bv - bv.mean()
    den = float(np.sqrt((av * av).sum() * (bv * bv).sum()))
    return float((av * bv).sum() / den) if den > 0 else None


def main():
    picks = json.loads(io.open('render-match.json', encoding='utf-8').read())
    tab = json.loads(io.open(os.path.join(OUT, 'idmap.json'), encoding='utf-8').read())
    names = tab['names'][:len(tab['colours'])]
    cols = [c for c in tab['colours'] if c]
    names = names[:len(cols)]
    print('THE DECISION RULE, WRITTEN OUT BEFORE THE SCORES ARE READ.')
    print('   A surface is read only where the model draws at least %d pixels of it after the mask is'
          % MINPIX)
    print('   eroded by %d, so no antialiased edge is counted. Its score is the gradient correlation'
          % ERODE)
    print('   inside that mask, taken where a PERFECT model would put the render. THE CONTROL IS THE SAME')
    print('   MASK ON A DIFFERENT FRAME PHOTOGRAPH, which answers the objection that a mask over a busy')
    print('   part of a picture scores well whatever is drawn there. CORROBORATED means the true score')
    print('   beats every control for that same mask. Anything else is NOT CORROBORATED, which is not')
    print('   refuted and will not be written as though it were.')
    print('')
    frames = []
    for i, p in enumerate(picks):
        rp, ip = os.path.join(OUT, 'r%02d.jpg' % i), os.path.join(OUT, 'id%02d.jpg' % i)
        if not (os.path.exists(rp) and os.path.exists(ip)):
            continue
        A = photo(p)
        rn = cv2.imread(rp, cv2.IMREAD_GRAYSCALE)
        idm = cv2.imread(ip, cv2.IMREAD_COLOR)
        if A is None or rn is None or idm is None:
            continue
        B = grad(turn(p, rn))
        ID = turn(p, idm, nearest=True)
        frames.append((i, p, A, B, ID))
    if not frames:
        sys.exit('   no rendered pair on disk, nothing to score')

    # HOW MUCH OF THE IDENTITY PASS SURVIVED THE TRIP, asked before anything is read from it
    tot = cls = 0
    for _, _, _, _, ID in frames:
        px = ID.reshape(-1, 3).astype(np.int16)
        near = np.full(len(px), 10 ** 6)
        for c in cols + [[0, 0, 0]]:
            d = np.abs(px - np.array([c[2], c[1], c[0]], np.int16)).max(1)
            near = np.minimum(near, d)
        tot += len(px)
        cls += int((near <= 28).sum())
    print('   THE IDENTITY PASS SURVIVED THE SCREENSHOT: %.1f per cent of pixels sit within 28 levels of'
          % (100.0 * cls / max(tot, 1)))
    print('   a colour that was actually assigned, over %d frames. Anything further off is an edge between'
          % len(frames))
    print('   two surfaces and is not counted for either of them.')
    if cls / max(tot, 1) < 0.9:
        print('')
        print('   THAT IS TOO LOW TO READ ANYTHING FROM. The colours did not come back intact, so the')
        print('   masks below would not be the surfaces they claim to be. NOTHING IS CONCLUDED.')
        sys.exit(0)

    ker = np.ones((ERODE, ERODE), np.uint8)
    per = {}
    for i, p, A, B, ID in frames:
        sc = WH / float(p['hgt'])
        dx, dy = int(round(p['ppx'] * sc)), int(round(p['ppy'] * sc))
        for nm, c in zip(names, cols):
            d = np.abs(ID.astype(np.int16) - np.array([c[2], c[1], c[0]], np.int16)).max(2)
            m = cv2.erode((d <= 28).astype(np.uint8), ker).astype(bool)
            if m.sum() < MINPIX:
                continue
            s = masked_ncc(A, B, m, dx, dy)
            if s is None:
                continue
            ctrl = []
            for j, q, A2, _, _ in frames:
                if j == i:
                    continue
                a2 = A2 if A2.shape == A.shape else cv2.resize(A2, (A.shape[1], A.shape[0]))
                v = masked_ncc(a2, B, m, dx, dy)
                if v is not None:
                    ctrl.append(v)
            if not ctrl:
                continue
            per.setdefault(nm, []).append((p['cls'], p['frame'], int(m.sum()), s, max(ctrl)))

    if not per:
        print('')
        print('   NOT ONE SURFACE COVERS ENOUGH OF ANY FRAME TO BE READ. Nothing is concluded.')
        sys.exit(0)

    print('')
    print('   surface                     frames  best score  its control  corroborated in')
    rank = []
    for nm, rows in per.items():
        bestrow = max(rows, key=lambda t: t[3] - t[4])
        good = sum(1 for t in rows if t[3] > t[4])
        rank.append((good, bestrow[3] - bestrow[4], nm, rows, bestrow, good))
    rank.sort(key=lambda t: (-t[0], -t[1]))
    for good, marg, nm, rows, bestrow, _ in rank:
        print('   %-26s %4d    %+.4f     %+.4f      %d of %d  %s'
              % (nm[:26], len(rows), bestrow[3], bestrow[4], good, len(rows),
                 bestrow[0] + ' ' + bestrow[1]))
    print('')
    corr = [t for t in rank if t[0] > 0]
    dead = [t for t in rank if t[0] == 0]
    print('   %d of the %d surfaces the goal names are CORROBORATED in at least one frame: the photograph'
          % (len(corr), len(rank)))
    print('   agrees with what the model draws there better than it agrees with any other frame.')
    if dead:
        print('')
        print('   AND %d ARE DRAWN ON NOTHING THESE FRAMES CAN SEE. That is a statement about the archive'
              % len(dead))
        print('   and not about the building: no frame in this set looks at them with enough pixels and')
        print('   enough contrast to tell. They are where the next measurement should go.')
        for good, marg, nm, rows, bestrow, _ in dead:
            print('      %-26s best %+.4f against a control of %+.4f, on %s %s'
                  % (nm[:26], bestrow[3], bestrow[4], bestrow[0], bestrow[1]))
    print('')
    print('   WHAT THIS CANNOT DO, said plainly. The mask comes from the MODEL, so a surface drawn in')
    print('   completely the wrong place is scored against whatever the photograph shows there and simply')
    print('   fails, without pointing anywhere. It cannot separate a wrong surface from a wrong pose. And')
    print('   the baked scan meshes came from photographs of this building, so where one of those scores')
    print('   well, part of the agreement is the bake agreeing with its own source.')


if __name__ == '__main__':
    main()
